torch_geometric/explain/algorithm/gnn_explainer.py

```diff
@@ -0,0 +1,539 @@
+        'EPS': 1e-15,
+    }
+
+    def __init__(
+        self,
+        epochs: int = 100,
+        lr: float = 0.01,
+        **kwargs,
+    ):
+        super().__init__()
+        self.epochs = epochs
+        self.lr = lr
+        self.coeffs.update(kwargs)
+        # node_mask will never actually be None
+        self.node_mask = None
+        # edge_mask is only None if edge_mask_type is != MaskType.object
+        self.edge_mask = None
+
+    def forward(
+        self,
+        model: torch.nn.Module,
+        x: Tensor,
+        edge_index: Tensor,
+        explainer_config: ExplainerConfig,
+        model_config: ModelConfig,
+        target: Tensor,
+        target_index: Optional[Union[int, Tensor]] = None,
+        node_index: Optional[int] = None,
+        **kwargs,
+    ) -> Explanation:
+
+        if isinstance(target_index, Tensor) and target_index.numel() > 1:
+            raise ValueError(
+                "GNNExplainer only supports single target index for now")
+
+        assert model_config.task_level in [
+            ModelTaskLevel.graph, ModelTaskLevel.node
+        ]
+
+        model.eval()
+
+        if model_config.task_level == ModelTaskLevel.node:
+            node_mask, edge_mask = self._explain_node(model, x, edge_index,
+                                                      explainer_config,
+                                                      model_config, target,
+                                                      node_index, target_index,
+                                                      **kwargs)
+        elif model_config.task_level == ModelTaskLevel.graph:
+            node_mask, edge_mask = self._explain_graph(model, x, edge_index,
+                                                       explainer_config,
+                                                       model_config, target,
+                                                       target_index, **kwargs)
+
+        if explainer_config.node_mask_type == MaskType.object:
+            node_feat_mask = None
+        else:
+            node_feat_mask = node_mask
+            node_mask = None
+
+        self._clean_model(model)
+
+        # build explanation
+        return Explanation(x=x, edge_index=edge_index, edge_mask=edge_mask,
+                           node_mask=node_mask, node_feat_mask=node_feat_mask)
+
+    def _explain_graph(
+        self,
+        model: torch.nn.Module,
+        x: Tensor,
+        edge_index: Tensor,
+        explainer_config: ExplainerConfig,
+        model_config: ModelConfig,
+        target: Tensor,
+        target_index: Optional[Union[int, Tensor]] = None,
+        **kwargs,
+    ) -> Tuple[Tensor, Optional[Tensor]]:
+        self._train_node_edge_mask(model, x, edge_index, explainer_config,
+                                   model_config, target, target_index, None,
+                                   **kwargs)
+
+        node_mask = self.node_mask.detach().sigmoid().squeeze(-1)
```

Comment:
Does this work for common_attributes? I also noticed in `check_explanation` in `test_gnn_explainer`, `MaskType.common_attributes` is not considered...

Additionally, can this and the mask transforms in `_explain_node()` be consolidated to prevent future errors.

Reply:
I was mainly wondering if _explain_graph() handles explainer_config.node_mask_type == MaskType.common_attributes. In _explain_node(), _reshape_common_attributes is called on the node_mask, but that check isn't present in _explain_graph. If that is correct, then I have no additional feedback.

If we do need to _reshape_common_attributes() in _explain_graph(), then there is a section of code in both _explain_graph and _explain_node that handles creating the correct node_mask and edge_mask, and this can be abstracted out to either forward() or another function called in forward().

Does that make sense? I can clarify further if needed.

Reply:
Oh yeah it make sens. It could but I'm not sure it's worth it, it might make some mechanism in the explain_node a bit less obvious (the part were the computational subgraph is selected, which requires some update on the node mask).